ethexe/contracts/src/ScaleCodec.sol

```diff
@@ -0,0 +1,508 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.26;
+
+function bytesToUint(bytes memory data, uint256 byteLength, uint256 offset) pure returns (uint256) {
+    uint256 result = 0;
+
+    assembly {
+        let src_ptr := add(add(data, 0x20), offset)
+        for { let i := 0 } lt(i, byteLength) { i := add(i, 1) } {
+            let byte_value := byte(0, mload(add(src_ptr, i)))
+            result := or(result, shl(mul(i, 8), byte_value))
+        }
+    }
+
+    return result;
+}
+
+library ScaleCodec {
+    struct CompactInt {
+        uint256 value;
+        uint256 offset;
+    }
+
+    struct DecodedString {
+        string value;
+        uint256 offset;
+    }
+
+    struct Optional {
+        bool isSome;
+        bytes value;
+    }
+
+    struct Result {
+        bool isOk;
+        bool isErr;
+        bytes value;
+    }
+
+    function sliceBytes(bytes memory data, uint256 start, uint256 end) public pure returns (bytes memory) {
+        bytes memory result = new bytes(end - start);
+
+        for (uint256 i = 0; i < end - start; i++) {
+            result[i] = data[i + start];
```

Comment:
why do you need to fill each slice/bytes etc?

Reply:
everywhere

Reply:
discussed